lib/sensu/client/process.rb

```diff
@@ -110,6 +110,46 @@ def check_in_progress_key(check)
         [check[:source], check[:name]].compact.join(":")
       end
 
+      # Execute a check hook, capturing its output (STDOUT/ERR),
+      # exit status code, executed timestamp, and duration. This
+      # method determines which hook command to run by inspecting the
+      # check execution status. Check hook command tokens are
+      # substituted with the associated client attribute values, via
+      # `substitute_tokens()`. If there are unmatched check attribute
+      # value tokens, the check hook will not be executed, instead the
+      # hook command output will be set to report the unmatched tokens.
+      #
+      # @param check [Hash]
+      # @yield [check] callback/block called after executing the check
+      #   hook (if any).
+      def execute_check_hook(check)
+        @logger.debug("attempting to execute check hook", :check => check)
+        severity = SEVERITIES[check[:status]] || "unknown"
+        hook = check[:hooks][check[:status].to_s] || check[:hooks][severity.to_sym]
+        if hook.nil? && check[:status] != 0
+          hook = check[:hooks]["non-zero"]
+        end
+        if hook
+          command, unmatched_tokens = substitute_tokens(hook[:command], @settings[:client])
```

Comment:
in addition to leveraging client config tokens, it could be convenient to provide/document a mechanism to pass all or parts of check result event to the hook's command as well.

Reply:
Oh, that's a good idea.

Reply:
Glad to hear 😄,

...  because another thought that I cannot quite shake is that this is new feature is bound to be used to cannabelize server side handlers/extensions. For example, an easy temptation would be to use those hooks to fanout processing logic that can get problematic when funneled through a reduced number of sensu server instances.

I fully understand that what follows is quite a stretch from the intent of this pull request but... 

In effect what we have here is a poor's man client side handler but without the supporting framework that server side handlers/extensions enjoy (configuration, logging, defined interface, etc.).

And if we follow this train of thought to it logical conclusion... should this hook concept really be about allowing the client to run handlers/extension to process check result payloads?

It would arguably be a heavier process than the proposed desgin (at the very least you would likely be bound to use ruby) but it would also be more powerfully integrated, more inlined with the existing patterns, would not require to specify a separate command/handler for each status value, etc...

I feel a bit like the devil's advocate here but still,... food for thoughts.

d.

Reply:
@davideauee thank you for the feedback!

I could add `"stdin": true` as a hook option (default to `false`), and have the Sensu client write the check definition with the populated `output`, `status`, etc to the hook process STDIN 👍 

Providing "semi event data" to the hook process STDIN is a possibility, however, it would be lacking the context constructed by the Sensu server, e.g. history, flapping, etc, and functionality like silencing would not work. I personally do not like the idea of overloading Check Hooks.

Reply:
leveraging STDIN would do the trick indeed 👍 .

I see what you mean by not wanting to "overload" Check Hooks as the intent here is to provide something both light weight and pragmatic, but it arguably comes at a cost of some amount of shoehorning.   

looking at the pros of each approach:
- client-side handlers:
   - more consistent with already established and well understood patterns 
   - well defined interfaces to feed data (configs, event data, etc.) so no need to shoehorn it through more esoteric means (config tokens, split triggers based on status, stdin, etc.)
   - could have its own config
   - better at promoting reuse
- check hooks:
   - not limited to ruby
   - easier to leverage existing local admin/ops tools (no need for a ruby shim) 
   - should be faster (at the very least, no need to bring up a ruby vm)